File: Richie/Word search/word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        if not board:
            return False
        
        
        
        word_length = len(word)
        h, w = len(board), len(board[0])
        
        
        def dfs(r,c,point=0):
            if point >= word_length:
                return True
            
            if r<0 or c<0 or c>= w or r>=h or board[r][c] != word[point]:
                return False
            char = board[r][c]
            board[r][c] = "#"
            
            result = dfs(r+1,c , point + 1) or dfs(r-1, c, point +1)  or dfs(r,c +1 , point + 1) or dfs(r, c-1, point +1)
            if result:
                return result
            board[r][c] = char
            return result 
        
        for r in range(h):
            for c in range(w):
                if board[r][c] == word[0]:
                    if dfs(r, c):
                        return True
        return False

File: Engr-T.stark/wordSearch.cpp
class Solution {
public:
    bool find=false;
    void dfs(vector<vector<bool>> &vis,vector<vector<char>>& board, string word,int i,int j,int c){
        if(c>=word.length()) {
            find=true;
                             return;}
        if( i<0 || j<0 || i>=board.size() || j>=board[0].size()|| vis[i][j] || find) return;
        vis[i][j]=true;
        
        if(word[c]==board[i][j]){
        dfs(vis,board,word,i+1,j,c+1);
            dfs(vis,board,word,i,j+1,c+1);
            dfs(vis,board,word,i-1,j,c+1);
            dfs(vis,board,word,i,j-1,c+1);
            }
    vis[i][j]=false;
    }
    bool exist(vector<vector<char>>& board, string word) {
           vector<vector<bool>> vis( board.size() , vector<bool> (board[0].size(),false)); 
        

        for(int i=0;i<board.size();i++){
            for(int j=0;j<board[0].size();j++){
                if(find==true) return true;
                if(!vis[i][j]){
                    dfs(vis,board,word,i,j,0);
                }
                
             
                
            }
        }
        return find;
    }
};
